filters/builtin/compress.go

```diff
@@ -0,0 +1,323 @@
+// applies the no-transform pragma, or the Content-Type is set to an unsupported
+// value.
+//
+// The default supported content types are: text/plain, text/html,
+// application/json, application/javascript, application/x-javascript,
+// text/javascript, text/css, image/svg+xml, application/octet-stream.
+//
+// The default set of MIME types can be reset or extended by passing in the desired
+// types as filter arguments. When extending the defaults, the first argument needs
+// to be "...". E.g. to compress tiff in addition to the defaults:
+//
+// 	* -> compress("...", "image/tiff") -> "https://www.example.org"
+//
+// To reset the supported types, e.g. to compress only HTML, the "..." argument
+// needs to be omitted:
+//
+// 	* -> compress("text/html") -> "https://www.example.org"
+//
+// The filter also checks the incoming request, if it accepts the supported
+// encodings, explicitly stated in the Accept-Encoding header. The filter currently
+// supports gzip and deflate. It does not assume that the client accepts any
+// encoding if the Accept-Encoding header is not set. It ignores * in the
+// Accept-Encoding header.
+//
+// When compressing the response, it updates the response header. It deletes the
+// the Content-Length value triggering the proxy to always return the response
+// with chunked transfer encoding, sets the Content-Encoding to the selected
+// encoding and sets the Vary: Accept-Encoding header, if missing.
+//
+// The compression happens in a streaming way, using only a small internal buffer.
+//
+func NewCompress() filters.Spec { return &compress{} }
+
+func (c *compress) Name() string {
+	return CompressName
+}
+
+func (c *compress) CreateFilter(args []interface{}) (filters.Filter, error) {
+	f := &compress{}
+
+	if len(args) == 0 {
+		f.mime = defaultCompressMIME
+		return f, nil
+	}
+
+	if args[0] == "..." {
+		f.mime = defaultCompressMIME
+		args = args[1:]
+	}
+
+	for _, a := range args {
+		if s, ok := a.(string); ok {
+			f.mime = append(f.mime, s)
+		} else {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+	}
+
+	return f, nil
+}
+
+func (c *compress) Request(_ filters.FilterContext) {}
+
+func stringsContain(ss []string, s string, transform ...func(string) string) bool {
+	for _, si := range ss {
+		for _, t := range transform {
+			si = t(si)
+		}
+
+		if si == s {
+			return true
+		}
+	}
+
+	return false
+}
+
+func canEncodeEntity(r *http.Response, mime []string) bool {
+	if ce := r.Header.Get("Content-Encoding"); ce != "" && ce != "identity" /* forgiving for identity */ {
+		return false
+	}
+
+	cc := strings.Split(r.Header.Get("Cache-Control"), ",")
+	if stringsContain(cc, "no-transform", strings.TrimSpace, strings.ToLower) {
+		return false
+	}
+
+	ct := r.Header.Get("Content-Type")
+	if i := strings.Index(ct, ";"); i >= 0 {
+		ct = ct[:i]
+	}
+
+	if !stringsContain(mime, ct) {
+		return false
+	}
+
+	return true
+}
+
+func acceptedEncoding(r *http.Request) string {
+	var encs encodings
```

Comment:
I'm kind of surprised that there is nothing like that in the go std lib... But the code is clear and understandable... so no offense